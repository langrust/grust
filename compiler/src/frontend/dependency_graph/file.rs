use petgraph::algo::toposort;
use petgraph::graphmap::DiGraphMap;

prelude! {
    error::{Error, TerminationError},
    hir::file::File,
    symbol_table::SymbolTable,

}

impl File {
    /// Generate dependency graph for every nodes/component.
    pub fn generate_dependency_graphs(
        &mut self,
        symbol_table: &SymbolTable,
        errors: &mut Vec<Error>,
    ) -> Result<(), TerminationError> {
        let File { nodes, .. } = self;

        // initialize dictionariy for reduced graphs
        let mut nodes_reduced_graphs = HashMap::new();

        // create graph of nodes
        let mut nodes_graph = DiGraphMap::new();
        nodes
            .iter()
            .for_each(|node| node.add_node_dependencies(&mut nodes_graph));

        // sort nodes according to their dependencies
        let sorted_nodes = toposort(&nodes_graph, None).map_err(|node| {
            let error = Error::NotCausalNode {
                node: symbol_table.get_name(node.node_id()).clone(),
                location: self.location.clone(),
            };
            errors.push(error);
            TerminationError
        })?;
        nodes.sort_by(|n1, n2| {
            let index1 = sorted_nodes
                .iter()
                .position(|id| *id == n1.id)
                .expect("should be in sorted list");
            let index2 = sorted_nodes
                .iter()
                .position(|id| *id == n2.id)
                .expect("should be in sorted list");

            Ord::cmp(&index2, &index1)
        });

        // ordered nodes complete their dependency graphs
        nodes
            .iter_mut()
            .map(|node| node.compute_dependencies(symbol_table, &mut nodes_reduced_graphs, errors))
            .collect::<Vec<Result<(), TerminationError>>>()
            .into_iter()
            .collect::<Result<(), TerminationError>>()?;

        Ok(())
    }
}
