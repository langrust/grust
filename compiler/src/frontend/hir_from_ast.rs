prelude! {}

/// HIR Component construction from AST Component
pub mod component;
/// HIR Contract construction from AST Contract
pub mod contract;
/// HIR Equation construction from AST Equation
pub mod equation;
/// HIR Expression construction from AST Expression
pub mod expression;
/// HIR File construction from AST File
pub mod file;
/// HIR Function construction from AST Function
pub mod function;
/// HIR Interface construction from AST Interface.
pub mod interface;
/// HIR Pattern construction from AST Pattern
pub mod pattern;
/// HIR Statement construction from AST Statement
pub mod statement;
/// HIR StreamExpression construction from AST StreamExpression
pub mod stream_expression;
pub mod typ;
/// HIR Typedef construction from AST Typedef.
pub mod typedef;

/// AST transformation into HIR.
pub trait HIRFromAST<Ctxt> {
    /// Corresponding HIR construct.
    type HIR;

    /// Transforms AST into HIR and check identifiers good use.
    fn hir_from_ast(self, ctxt: &mut Ctxt) -> TRes<Self::HIR>;
}

pub struct SimpleCtxt<'a> {
    pub syms: &'a mut SymbolTable,
    pub errors: &'a mut Vec<Error>,
}
pub struct LocCtxt<'a> {
    pub loc: &'a Location,
    pub syms: &'a mut SymbolTable,
    pub errors: &'a mut Vec<Error>,
}
pub struct PatLocCtxt<'a> {
    pub pat: Option<&'a ast::stmt::Pattern>,
    pub loc: &'a Location,
    pub syms: &'a mut SymbolTable,
    pub errors: &'a mut Vec<Error>,
}
impl<'a> SimpleCtxt<'a> {
    pub fn new(syms: &'a mut SymbolTable, errors: &'a mut Vec<Error>) -> Self {
        Self { syms, errors }
    }
    pub fn add_loc<'b>(&'b mut self, loc: &'b Location) -> LocCtxt<'b> {
        LocCtxt::new(loc, self.syms, self.errors)
    }
    pub fn add_pat_loc<'b>(
        &'b mut self,
        pat: Option<&'b ast::stmt::Pattern>,
        loc: &'b Location,
    ) -> PatLocCtxt<'b> {
        PatLocCtxt::new(pat, loc, self.syms, self.errors)
    }
}
impl<'a> LocCtxt<'a> {
    pub fn new(loc: &'a Location, syms: &'a mut SymbolTable, errors: &'a mut Vec<Error>) -> Self {
        Self { loc, syms, errors }
    }
    pub fn add_pat<'b>(&'b mut self, pat: Option<&'b ast::stmt::Pattern>) -> PatLocCtxt<'b> {
        PatLocCtxt::new(pat, self.loc, self.syms, self.errors)
    }
}
impl<'a> PatLocCtxt<'a> {
    pub fn new(
        pat: Option<&'a ast::stmt::Pattern>,
        loc: &'a Location,
        syms: &'a mut SymbolTable,
        errors: &'a mut Vec<Error>,
    ) -> Self {
        Self {
            pat,
            loc,
            syms,
            errors,
        }
    }
    pub fn remove_pat<'b>(&'b mut self) -> LocCtxt<'b> {
        LocCtxt::new(self.loc, self.syms, self.errors)
    }
    pub fn set_pat(
        &mut self,
        pat: Option<&'a ast::stmt::Pattern>,
    ) -> Option<&'a ast::stmt::Pattern> {
        std::mem::replace(&mut self.pat, pat)
    }
}
