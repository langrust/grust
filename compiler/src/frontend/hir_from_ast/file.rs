prelude! {
    ast::Ast,
    hir::interface::Interface,
}

use super::HIRFromAST;

impl HIRFromAST for Ast {
    type HIR = hir::File;

    fn hir_from_ast(
        self,
        symbol_table: &mut SymbolTable,
        errors: &mut Vec<Error>,
    ) -> TRes<Self::HIR> {
        // initialize symbol table with builtin operators
        symbol_table.initialize();

        // store elements in symbol table
        self.store(symbol_table, errors)?;

        let Ast { items } = self;

        let (typedefs, functions, components, imports, exports, services) = items.into_iter().fold(
            (vec![], vec![], vec![], vec![], vec![], vec![]),
            |(
                mut typedefs,
                mut functions,
                mut components,
                mut imports,
                mut exports,
                mut services,
            ),
             item| {
                match item {
                    ast::Item::Component(component) => {
                        components.push(component.hir_from_ast(symbol_table, errors))
                    }
                    ast::Item::Function(function) => {
                        functions.push(function.hir_from_ast(symbol_table, errors))
                    }
                    ast::Item::Typedef(typedef) => {
                        typedefs.push(typedef.hir_from_ast(symbol_table, errors))
                    }
                    ast::Item::Service(service) => {
                        services.push(service.hir_from_ast(symbol_table, errors))
                    }
                    ast::Item::Import(import) => imports.push(
                        import
                            .hir_from_ast(symbol_table, errors)
                            .map(|res| (symbol_table.get_fresh_id(), res)),
                    ),
                    ast::Item::Export(export) => exports.push(
                        export
                            .hir_from_ast(symbol_table, errors)
                            .map(|res| (symbol_table.get_fresh_id(), res)),
                    ),
                    ast::Item::ComponentImport(component) => {
                        components.push(component.hir_from_ast(symbol_table, errors))
                    }
                }
                (typedefs, functions, components, imports, exports, services)
            },
        );

        let interface = Interface {
            services: services.into_iter().collect::<TRes<Vec<_>>>()?,
            imports: imports.into_iter().collect::<TRes<_>>()?,
            exports: exports.into_iter().collect::<TRes<_>>()?,
        };

        Ok(hir::File {
            typedefs: typedefs.into_iter().collect::<TRes<Vec<_>>>()?,
            functions: functions.into_iter().collect::<TRes<Vec<_>>>()?,
            components: components.into_iter().collect::<TRes<Vec<_>>>()?,
            interface,
            location: Location::default(),
        })
    }
}
