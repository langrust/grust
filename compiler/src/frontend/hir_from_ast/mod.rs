prelude! {}

/// HIR Contract construction from AST Contract
pub mod contract;
/// HIR Equation construction from AST Equation
pub mod equation;
/// HIR Expression construction from AST Expression
pub mod expression;
/// HIR File construction from AST File
pub mod file;
/// HIR Function construction from AST Function
pub mod function;
/// HIR Interface construction from AST Interface.
pub mod interface;
/// HIR Pattern construction from AST Pattern
pub mod pattern;
/// HIR Statement construction from AST Statement
pub mod statement;
/// HIR StreamExpression construction from AST StreamExpression
pub mod stream_expression;
/// HIR Typedef construction from AST Typedef.
pub mod typedef;

/// AST transformation into HIR.
pub trait HIRFromAST {
    /// Corresponding HIR construct.
    type HIR;

    /// Transforms AST into HIR and check identifiers good use.
    fn hir_from_ast(
        self,
        symbol_table: &mut SymbolTable,
        errors: &mut Vec<Error>,
    ) -> TRes<Self::HIR>;
}

impl HIRFromAST for ast::Component {
    type HIR = hir::Node;

    // precondition: node and its signals are already stored in symbol table
    // postcondition: construct HIR node and check identifiers good use
    fn hir_from_ast(
        self,
        symbol_table: &mut SymbolTable,
        errors: &mut Vec<Error>,
    ) -> TRes<Self::HIR> {
        let ast::Component {
            ident,
            contract,
            equations,
            ..
        } = self;
        let name = ident.to_string();
        let location = Location::default();
        let id = symbol_table.get_node_id(&name, false, location.clone(), errors)?;

        // create local context with all signals
        symbol_table.local();
        symbol_table.restore_context(id);
        symbol_table.enter_in_node(id);

        let statements = equations
            .into_iter()
            .map(|equation| equation.hir_from_ast(symbol_table, errors))
            .collect::<TRes<Vec<_>>>()?;
        let contract = contract.hir_from_ast(symbol_table, errors)?;

        symbol_table.leave_node();
        symbol_table.global();

        Ok(hir::Node {
            id,
            statements,
            contract,
            location,
            graph: graph::DiGraphMap::new(),
            memory: hir::Memory::new(),
        })
    }
}
