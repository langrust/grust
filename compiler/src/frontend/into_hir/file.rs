prelude! {
    ast::Ast,
    hir::interface::Interface,
}

impl IntoHir<hir::ctx::Simple<'_>> for Ast {
    type Hir = hir::File;

    fn into_hir(self, ctxt: &mut hir::ctx::Simple) -> TRes<Self::Hir> {
        // initialize symbol table with builtin operators
        ctxt.syms.initialize();

        // store elements in symbol table
        self.store(ctxt.syms, ctxt.errors)?;

        let Ast { items } = self;

        let (typedefs, functions, components, imports, exports, services) = items.into_iter().fold(
            (vec![], vec![], vec![], vec![], vec![], vec![]),
            |(
                mut typedefs,
                mut functions,
                mut components,
                mut imports,
                mut exports,
                mut services,
            ),
             item| {
                match item {
                    ast::Item::Component(component) => components.push(component.into_hir(ctxt)),
                    ast::Item::Function(function) => functions.push(function.into_hir(ctxt)),
                    ast::Item::Typedef(typedef) => typedefs.push(typedef.into_hir(ctxt)),
                    ast::Item::Service(service) => services.push(service.into_hir(ctxt)),
                    ast::Item::Import(import) => imports.push(
                        import
                            .into_hir(ctxt)
                            .map(|res| (ctxt.syms.get_fresh_id(), res)),
                    ),
                    ast::Item::Export(export) => exports.push(
                        export
                            .into_hir(ctxt)
                            .map(|res| (ctxt.syms.get_fresh_id(), res)),
                    ),
                    ast::Item::ComponentImport(component) => {
                        components.push(component.into_hir(ctxt))
                    }
                }
                (typedefs, functions, components, imports, exports, services)
            },
        );

        let interface = Interface {
            services: services.into_iter().collect::<TRes<Vec<_>>>()?,
            imports: imports.into_iter().collect::<TRes<_>>()?,
            exports: exports.into_iter().collect::<TRes<_>>()?,
        };

        Ok(hir::File {
            typedefs: typedefs.into_iter().collect::<TRes<Vec<_>>>()?,
            functions: functions.into_iter().collect::<TRes<Vec<_>>>()?,
            components: components.into_iter().collect::<TRes<Vec<_>>>()?,
            interface,
            location: Location::default(),
        })
    }
}
