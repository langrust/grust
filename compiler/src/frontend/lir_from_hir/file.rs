use crate::{
    hir::file::File,
    lir::{item::Item, project::Project},
    symbol_table::SymbolTable,
};

use super::LIRFromHIR;

impl LIRFromHIR for File {
    type LIR = Project;

    fn lir_from_hir(self, symbol_table: &SymbolTable) -> Self::LIR {
        let File {
            typedefs,
            functions,
            nodes,
            interface,
            ..
        } = self;

        let mut typedefs = typedefs
            .into_iter()
            .map(|typedef| typedef.lir_from_hir(symbol_table))
            .collect();
        let mut functions = functions
            .into_iter()
            .map(|function| function.lir_from_hir(symbol_table))
            .map(Item::Function)
            .collect();
        let mut state_machines = nodes
            .into_iter()
            .map(|node| node.lir_from_hir(symbol_table))
            .map(Item::StateMachine)
            .collect();
        let execution_machine = interface.lir_from_hir(symbol_table);

        let mut items = vec![];
        items.append(&mut typedefs);
        items.append(&mut functions);
        items.append(&mut state_machines);
        items.push(Item::ExecutionMachine(execution_machine));

        Project { items }
    }
}
