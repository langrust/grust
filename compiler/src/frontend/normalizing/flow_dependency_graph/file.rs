prelude! { just
    hir::File,
}

impl File {
    /// Generate dependency graphs for the interface.
    #[inline]
    pub fn generate_flows_dependency_graphs(&mut self) {
        self.interface.generate_flows_dependency_graphs()
    }
}
