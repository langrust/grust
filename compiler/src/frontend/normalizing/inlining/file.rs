use std::collections::HashMap;

use crate::{hir::file::File, symbol_table::SymbolTable};

impl File {
    /// Inline node application when it is needed.
    ///
    /// Inlining needed for "shifted causality loop".
    ///
    /// # Example:
    /// ```GR
    /// node semi_fib(i: int) {
    ///     out o: int = 0 fby (i + 1 fby i);
    /// }
    /// node fib_call() {
    ///    out fib: int = semi_fib(fib).o;
    /// }
    /// ```
    /// In this example, `fib_call` calls `semi_fib` with the same input and output signal.
    /// There is no causality loop, `o` depends on the memory of `i`.
    ///
    /// We need to inline the code, the output `fib` is defined before the input `fib`,
    /// which can not be computed by a function call.
    pub fn inline_when_needed(&mut self, symbol_table: &mut SymbolTable) {
        let nodes = self
            .nodes
            .iter()
            .flat_map(|node| {
                node.unitary_nodes
                    .values()
                    .map(|unitary_node| (unitary_node.id.clone(), unitary_node.clone()))
            })
            .collect::<HashMap<_, _>>();
        self.nodes
            .iter_mut()
            .for_each(|node| node.inline_when_needed(&nodes, symbol_table))
    }
}
