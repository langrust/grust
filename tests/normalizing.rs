use codespan_reporting::files::{Files, SimpleFiles};

use grustine::ast::file::File;
use grustine::frontend::hir_from_ast::file::hir_from_ast;
use grustine::parser::langrust;

#[test]
fn normalize_counter() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let counter_id = files.add(
        "counter.gr",
        std::fs::read_to_string("tests/fixture/counter.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(counter_id, &files.source(counter_id).unwrap())
        .unwrap();
    file.typing(&mut errors).unwrap();
    let mut file = hir_from_ast(file);
    file.generate_dependency_graphs(&mut errors).unwrap();
    file.causality_analysis(&mut errors).unwrap();

    file.normalize(&mut errors).unwrap();
    insta::assert_yaml_snapshot!(file);
}

#[test]
fn normalize_blinking() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let blinking_id = files.add(
        "blinking.gr",
        std::fs::read_to_string("tests/fixture/blinking.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(blinking_id, &files.source(blinking_id).unwrap())
        .unwrap();
    file.typing(&mut errors).unwrap();
    let mut file = hir_from_ast(file);
    file.generate_dependency_graphs(&mut errors).unwrap();
    file.causality_analysis(&mut errors).unwrap();

    file.normalize(&mut errors).unwrap();
    insta::assert_yaml_snapshot!(file);
}

#[test]
fn normalize_button_management() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let blinking_id = files.add(
        "button_management.gr",
        std::fs::read_to_string("tests/fixture/button_management.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(blinking_id, &files.source(blinking_id).unwrap())
        .unwrap();
    file.typing(&mut errors).unwrap();
    let mut file = hir_from_ast(file);
    file.generate_dependency_graphs(&mut errors).unwrap();
    file.causality_analysis(&mut errors).unwrap();

    file.normalize(&mut errors).unwrap();
    insta::assert_yaml_snapshot!(file);
}

#[test]
fn normalize_button_management_condition_match() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let blinking_id = files.add(
        "button_management_condition_match.gr",
        std::fs::read_to_string("tests/fixture/button_management_condition_match.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(blinking_id, &files.source(blinking_id).unwrap())
        .unwrap();
    file.typing(&mut errors).unwrap();
    let mut file = hir_from_ast(file);
    file.generate_dependency_graphs(&mut errors).unwrap();
    file.causality_analysis(&mut errors).unwrap();

    file.normalize(&mut errors).unwrap();
    insta::assert_yaml_snapshot!(file);
}

#[test]
fn normalize_button_management_using_function() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let blinking_id = files.add(
        "button_management_using_function.gr",
        std::fs::read_to_string("tests/fixture/button_management_using_function.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(blinking_id, &files.source(blinking_id).unwrap())
        .unwrap();
    file.typing(&mut errors).unwrap();
    let mut file = hir_from_ast(file);
    file.generate_dependency_graphs(&mut errors).unwrap();
    file.causality_analysis(&mut errors).unwrap();

    file.normalize(&mut errors).unwrap();
    insta::assert_yaml_snapshot!(file);
}
