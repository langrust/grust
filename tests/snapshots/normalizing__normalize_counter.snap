---
source: tests/normalizing.rs
expression: file
---
typedefs: []
functions: []
nodes:
  - id: 21
    unscheduled_equations:
      19:
        id: 19
        expression:
          kind:
            Expression:
              expression:
                Application:
                  function_expression:
                    kind:
                      Expression:
                        expression:
                          Identifier:
                            id: 15
                    typing:
                      Abstract:
                        - - Boolean
                          - Integer
                          - Integer
                        - Integer
                    location: ~
                    dependencies: []
                  inputs:
                    - kind:
                        Expression:
                          expression:
                            Identifier:
                              id: 17
                      typing: Boolean
                      location: ~
                      dependencies:
                        - - 17
                          - Weight: 0
                    - kind:
                        Expression:
                          expression:
                            Constant:
                              constant:
                                Integer: 0
                      typing: Integer
                      location: ~
                      dependencies: []
                    - kind:
                        Expression:
                          expression:
                            Application:
                              function_expression:
                                kind:
                                  Expression:
                                    expression:
                                      Identifier:
                                        id: 5
                                typing:
                                  Abstract:
                                    - - Integer
                                      - Integer
                                    - Integer
                                location: ~
                                dependencies: []
                              inputs:
                                - kind:
                                    Expression:
                                      expression:
                                        Application:
                                          function_expression:
                                            kind:
                                              Expression:
                                                expression:
                                                  Identifier:
                                                    id: 2
                                            typing:
                                              Abstract:
                                                - - Integer
                                                - Integer
                                            location: ~
                                            dependencies: []
                                          inputs:
                                            - kind:
                                                FollowedBy:
                                                  constant:
                                                    kind:
                                                      Expression:
                                                        expression:
                                                          Constant:
                                                            constant:
                                                              Integer: 0
                                                    typing: Integer
                                                    location: ~
                                                    dependencies: []
                                                  expression:
                                                    kind:
                                                      Expression:
                                                        expression:
                                                          Identifier:
                                                            id: 19
                                                    typing: Integer
                                                    location: ~
                                                    dependencies:
                                                      - - 19
                                                        - Weight: 0
                                              typing: Integer
                                              location: ~
                                              dependencies:
                                                - - 19
                                                  - Weight: 1
                                  typing: Integer
                                  location: ~
                                  dependencies:
                                    - - 19
                                      - Weight: 1
                                - kind:
                                    Expression:
                                      expression:
                                        Identifier:
                                          id: 20
                                  typing: Integer
                                  location: ~
                                  dependencies:
                                    - - 20
                                      - Weight: 0
                      typing: Integer
                      location: ~
                      dependencies:
                        - - 19
                          - Weight: 1
                        - - 20
                          - Weight: 0
          typing: Integer
          location: ~
          dependencies:
            - - 17
              - Weight: 0
            - - 19
              - Weight: 1
            - - 20
              - Weight: 0
        location: ~
      20:
        id: 20
        expression:
          kind:
            Expression:
              expression:
                Application:
                  function_expression:
                    kind:
                      Expression:
                        expression:
                          Identifier:
                            id: 15
                    typing:
                      Abstract:
                        - - Boolean
                          - Integer
                          - Integer
                        - Integer
                    location: ~
                    dependencies: []
                  inputs:
                    - kind:
                        Expression:
                          expression:
                            Identifier:
                              id: 18
                      typing: Boolean
                      location: ~
                      dependencies:
                        - - 18
                          - Weight: 0
                    - kind:
                        Expression:
                          expression:
                            Constant:
                              constant:
                                Integer: 1
                      typing: Integer
                      location: ~
                      dependencies: []
                    - kind:
                        Expression:
                          expression:
                            Constant:
                              constant:
                                Integer: 0
                      typing: Integer
                      location: ~
                      dependencies: []
          typing: Integer
          location: ~
          dependencies:
            - - 18
              - Weight: 0
        location: ~
    unitary_nodes:
      19:
        id: 25
        statements:
          - id: 20
            expression:
              kind:
                Expression:
                  expression:
                    Application:
                      function_expression:
                        kind:
                          Expression:
                            expression:
                              Identifier:
                                id: 15
                        typing:
                          Abstract:
                            - - Boolean
                              - Integer
                              - Integer
                            - Integer
                        location: ~
                        dependencies: []
                      inputs:
                        - kind:
                            Expression:
                              expression:
                                Identifier:
                                  id: 18
                          typing: Boolean
                          location: ~
                          dependencies:
                            - - 18
                              - Weight: 0
                        - kind:
                            Expression:
                              expression:
                                Constant:
                                  constant:
                                    Integer: 1
                          typing: Integer
                          location: ~
                          dependencies: []
                        - kind:
                            Expression:
                              expression:
                                Constant:
                                  constant:
                                    Integer: 0
                          typing: Integer
                          location: ~
                          dependencies: []
              typing: Integer
              location: ~
              dependencies:
                - - 18
                  - Weight: 0
            location: ~
          - id: 19
            expression:
              kind:
                Expression:
                  expression:
                    Application:
                      function_expression:
                        kind:
                          Expression:
                            expression:
                              Identifier:
                                id: 15
                        typing:
                          Abstract:
                            - - Boolean
                              - Integer
                              - Integer
                            - Integer
                        location: ~
                        dependencies: []
                      inputs:
                        - kind:
                            Expression:
                              expression:
                                Identifier:
                                  id: 17
                          typing: Boolean
                          location: ~
                          dependencies:
                            - - 17
                              - Weight: 0
                        - kind:
                            Expression:
                              expression:
                                Constant:
                                  constant:
                                    Integer: 0
                          typing: Integer
                          location: ~
                          dependencies: []
                        - kind:
                            Expression:
                              expression:
                                Application:
                                  function_expression:
                                    kind:
                                      Expression:
                                        expression:
                                          Identifier:
                                            id: 5
                                    typing:
                                      Abstract:
                                        - - Integer
                                          - Integer
                                        - Integer
                                    location: ~
                                    dependencies: []
                                  inputs:
                                    - kind:
                                        Expression:
                                          expression:
                                            Application:
                                              function_expression:
                                                kind:
                                                  Expression:
                                                    expression:
                                                      Identifier:
                                                        id: 2
                                                typing:
                                                  Abstract:
                                                    - - Integer
                                                    - Integer
                                                location: ~
                                                dependencies: []
                                              inputs:
                                                - kind:
                                                    Expression:
                                                      expression:
                                                        Identifier:
                                                          id: 28
                                                  typing: Integer
                                                  location: ~
                                                  dependencies:
                                                    - - 28
                                                      - Weight: 0
                                      typing: Integer
                                      location: ~
                                      dependencies:
                                        - - 19
                                          - Weight: 1
                                    - kind:
                                        Expression:
                                          expression:
                                            Identifier:
                                              id: 20
                                      typing: Integer
                                      location: ~
                                      dependencies:
                                        - - 20
                                          - Weight: 0
                          typing: Integer
                          location: ~
                          dependencies:
                            - - 19
                              - Weight: 1
                            - - 20
                              - Weight: 0
              typing: Integer
              location: ~
              dependencies:
                - - 17
                  - Weight: 0
                - - 19
                  - Weight: 1
                - - 20
                  - Weight: 0
            location: ~
        memory:
          buffers:
            28:
              typing: Integer
              initial_expression:
                kind:
                  Expression:
                    expression:
                      Constant:
                        constant:
                          Integer: 0
                typing: Integer
                location: ~
                dependencies: []
              expression:
                kind:
                  Expression:
                    expression:
                      Identifier:
                        id: 19
                typing: Integer
                location: ~
                dependencies:
                  - - 19
                    - Weight: 0
          called_nodes: {}
        location: ~
        graph:
          nodes:
            - 19
            - 20
            - 28
            - 17
            - 18
          node_holes: []
          edge_property: directed
          edges:
            - - 0
              - 3
              - Weight: 0
            - - 0
              - 0
              - Weight: 1
            - - 0
              - 1
              - Weight: 0
            - - 1
              - 4
              - Weight: 0
        contract:
          requires: []
          ensures: []
          invariant: []
    contract:
      requires: []
      ensures: []
      invariant: []
    location: ~
    graph:
      ? - 19
        - 17
      : Weight: 0
      ? - 19
        - 19
      : Weight: 1
      ? - 19
        - 20
      : Weight: 0
      ? - 20
        - 18
      : Weight: 0
component:
  id: 24
  unscheduled_equations:
    22:
      id: 22
      expression:
        kind:
          NodeApplication:
            node_id: 21
            inputs:
              - - 17
                - kind:
                    FollowedBy:
                      constant:
                        kind:
                          Expression:
                            expression:
                              Constant:
                                constant:
                                  Boolean: false
                        typing: Boolean
                        location: ~
                        dependencies: []
                      expression:
                        kind:
                          Expression:
                            expression:
                              Application:
                                function_expression:
                                  kind:
                                    Expression:
                                      expression:
                                        Identifier:
                                          id: 2
                                  typing:
                                    Abstract:
                                      - - Boolean
                                      - Boolean
                                  location: ~
                                  dependencies: []
                                inputs:
                                  - kind:
                                      Expression:
                                        expression:
                                          Application:
                                            function_expression:
                                              kind:
                                                Expression:
                                                  expression:
                                                    Identifier:
                                                      id: 13
                                              typing:
                                                Abstract:
                                                  - - Integer
                                                    - Integer
                                                  - Boolean
                                              location: ~
                                              dependencies: []
                                            inputs:
                                              - kind:
                                                  Expression:
                                                    expression:
                                                      Identifier:
                                                        id: 22
                                                typing: Integer
                                                location: ~
                                                dependencies:
                                                  - - 22
                                                    - Weight: 0
                                              - kind:
                                                  Expression:
                                                    expression:
                                                      Constant:
                                                        constant:
                                                          Integer: 35
                                                typing: Integer
                                                location: ~
                                                dependencies: []
                                    typing: Boolean
                                    location: ~
                                    dependencies:
                                      - - 22
                                        - Weight: 0
                        typing: Boolean
                        location: ~
                        dependencies:
                          - - 22
                            - Weight: 0
                  typing: Boolean
                  location: ~
                  dependencies:
                    - - 22
                      - Weight: 1
              - - 18
                - kind:
                    Expression:
                      expression:
                        Identifier:
                          id: 23
                  typing: Boolean
                  location: ~
                  dependencies:
                    - - 23
                      - Weight: 0
            output_id: 19
        typing: Integer
        location: ~
        dependencies:
          - - 22
            - Weight: 1
          - - 23
            - Weight: 0
      location: ~
    23:
      id: 23
      expression:
        kind:
          FollowedBy:
            constant:
              kind:
                Expression:
                  expression:
                    Constant:
                      constant:
                        Boolean: true
              typing: Boolean
              location: ~
              dependencies: []
            expression:
              kind:
                Expression:
                  expression:
                    Application:
                      function_expression:
                        kind:
                          Expression:
                            expression:
                              Identifier:
                                id: 1
                        typing:
                          Abstract:
                            - - Boolean
                            - Boolean
                        location: ~
                        dependencies: []
                      inputs:
                        - kind:
                            Expression:
                              expression:
                                Identifier:
                                  id: 23
                          typing: Boolean
                          location: ~
                          dependencies:
                            - - 23
                              - Weight: 0
              typing: Boolean
              location: ~
              dependencies:
                - - 23
                  - Weight: 0
        typing: Boolean
        location: ~
        dependencies:
          - - 23
            - Weight: 1
      location: ~
  unitary_nodes:
    22:
      id: 26
      statements:
        - id: 27
          expression:
            kind:
              Expression:
                expression:
                  Identifier:
                    id: 29
            typing: Boolean
            location: ~
            dependencies:
              - - 29
                - Weight: 0
          location: ~
        - id: 23
          expression:
            kind:
              Expression:
                expression:
                  Identifier:
                    id: 31
            typing: Boolean
            location: ~
            dependencies:
              - - 31
                - Weight: 0
          location: ~
        - id: 22
          expression:
            kind:
              UnitaryNodeApplication:
                node_id: 25
                inputs:
                  - - 17
                    - kind:
                        Expression:
                          expression:
                            Identifier:
                              id: 27
                      typing: Boolean
                      location: ~
                      dependencies:
                        - - 27
                          - Weight: 0
                  - - 18
                    - kind:
                        Expression:
                          expression:
                            Identifier:
                              id: 23
                      typing: Boolean
                      location: ~
                      dependencies:
                        - - 23
                          - Weight: 0
                output_id: 19
            typing: Integer
            location: ~
            dependencies:
              - - 27
                - Weight: 0
              - - 23
                - Weight: 0
          location: ~
      memory:
        buffers:
          29:
            typing: Boolean
            initial_expression:
              kind:
                Expression:
                  expression:
                    Constant:
                      constant:
                        Boolean: false
              typing: Boolean
              location: ~
              dependencies: []
            expression:
              kind:
                Expression:
                  expression:
                    Application:
                      function_expression:
                        kind:
                          Expression:
                            expression:
                              Identifier:
                                id: 2
                        typing:
                          Abstract:
                            - - Boolean
                            - Boolean
                        location: ~
                        dependencies: []
                      inputs:
                        - kind:
                            Expression:
                              expression:
                                Application:
                                  function_expression:
                                    kind:
                                      Expression:
                                        expression:
                                          Identifier:
                                            id: 13
                                    typing:
                                      Abstract:
                                        - - Integer
                                          - Integer
                                        - Boolean
                                    location: ~
                                    dependencies: []
                                  inputs:
                                    - kind:
                                        Expression:
                                          expression:
                                            Identifier:
                                              id: 22
                                      typing: Integer
                                      location: ~
                                      dependencies:
                                        - - 22
                                          - Weight: 0
                                    - kind:
                                        Expression:
                                          expression:
                                            Constant:
                                              constant:
                                                Integer: 35
                                      typing: Integer
                                      location: ~
                                      dependencies: []
                          typing: Boolean
                          location: ~
                          dependencies:
                            - - 22
                              - Weight: 0
              typing: Boolean
              location: ~
              dependencies:
                - - 22
                  - Weight: 0
          31:
            typing: Boolean
            initial_expression:
              kind:
                Expression:
                  expression:
                    Constant:
                      constant:
                        Boolean: true
              typing: Boolean
              location: ~
              dependencies: []
            expression:
              kind:
                Expression:
                  expression:
                    Application:
                      function_expression:
                        kind:
                          Expression:
                            expression:
                              Identifier:
                                id: 1
                        typing:
                          Abstract:
                            - - Boolean
                            - Boolean
                        location: ~
                        dependencies: []
                      inputs:
                        - kind:
                            Expression:
                              expression:
                                Identifier:
                                  id: 23
                          typing: Boolean
                          location: ~
                          dependencies:
                            - - 23
                              - Weight: 0
              typing: Boolean
              location: ~
              dependencies:
                - - 23
                  - Weight: 0
        called_nodes:
          30:
            node_id: 25
            signal_id: 19
      location: ~
      graph:
        nodes:
          - 27
          - 22
          - 23
          - 29
          - 31
        node_holes: []
        edge_property: directed
        edges:
          - - 0
            - 3
            - Weight: 0
          - - 1
            - 0
            - Weight: 0
          - - 1
            - 2
            - Weight: 0
          - - 2
            - 4
            - Weight: 0
      contract:
        requires: []
        ensures: []
        invariant: []
  contract:
    requires: []
    ensures: []
    invariant: []
  location: ~
  graph:
    ? - 22
      - 22
    : Weight: 1
    ? - 22
      - 23
    : Weight: 0
    ? - 23
      - 23
    : Weight: 1
location: ~
